rfcs/0107-version-normalization.md

```diff
@@ -0,0 +1,211 @@
+  - Unlabeled snapshot denotes any snapshot not explicitly labeled by its team.
+- Release denotes any distributed snapshot, as defined by its team.
+- Branch denotes a logical sequence of snapshots, as identified by the program's
+  team;
+  - Usually these branches are denoted by names such as `stable`, `master`,
+    `unstable`, `trunk`, `experimental`, `staging`, `X.Y` (where `X` and
+    `Y` are numbers) etc.
+
+## Design
+
+- For a labeled snapshot:
+  - `version` should be constituted of the version of the snapshot, as defined
+    by the program team, without any alphabetical characters (e.g. "v", "rel")
+    prepending it.
+    - Alphabetical characters following the first numerical character of version
+      (as defined above) should be maintained, except optionally those clearly
+      used as separators, in which case they are replaced by dots (emulating a
+      typical dot-separated version).
+
+- For an unlabeled snapshot:
+  - `version` should be constituted of a concatenation of the elements below in
+    this order:
+      - the version of the latest previous labeled snapshot (on the same branch,
+        when applicable), according to the rules defined before for labeled
+        snapshots;
+        - If the project never released a labeled snapshot, `0.pre` should be
+          used as default.
+     - the string `+date=YYYY-MM-DD`, where `YYYY-MM-DD` denotes the date
+       the mentioned unlabeled snapshot was distributed.
+
+- Some atypical programs have special considerations:
+  - Linux kernel modules:
+    - Besides the rules established above for typical snapshots (whether labeled
+      or unlabeled), `version` shoud have appended `+linux=XX.XX.XX`, where
+      `XX.XX.XX` is the corresponding Linux kernel version used to build the
+      aforementioned module.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Some useful examples:
+
+- Bochs is a typical Sourceforge-hosted project; its labeled snapshots can be
+  fetched from tarballs obtained via URLs like
+  <https://sourceforge.net/projects/bochs/files/bochs/2.6.11/>
+
+  For this example, we have `pname = "bochs"; version = "2.6.11";`.
+
+- MPV is a typical Github-hosted program; its labeled snapshots can be fetched
+  from tarballs obtained via URLs like
+  <https://github.com/mpv-player/mpv/releases/tag/v0.33.1>.
+
+  For this example, we get rid of the `"v"` prepended to the version tag: `pname
+  = "mpv"; version = "0.33.1";`.
+
+- SDL2 is hosted on Github; its latest labeled version can be downloaded from
+  <https://github.com/libsdl-org/SDL/releases/tag/release-2.0.14>. Therefore we
+  have `pname = "SDL2"; version = "2.0.14";`.
+
+  _However_, this labeled version was released December 21, 2020, while the
+  latest change was done in May 28, 2021.
+
+  Therefore, for this particular unlabeled releases of SDL2, we have `pname =
+  "SDL2"; version = "2.0.14+date=2021-05-28";`.
```

Comment:
Now we have a problem. We currently do not hold this information of what was the latest release. While I really appreciate the extra information which is very valuable for a reviewer I don't see how we want to convert the current scheme into this.

If we take a look how the pname+version RFC went: we are still converting packages in mass today and there are often people which insist on not setting version because something is not versioned.

What is the plan to convert from the current situation? How do we want to automate this treewide? If we don't have a plan I don't think we should move forward with this RFC because only maintained software will get updated and then a giant portion is left behind in some unmaintained limbo state.

Reply:
There is no commitment to the latest release here. I am using them as examples, not as rules.

If, by any reason, someone desired to compile and install a commit made in 2010-12-20, the rules would still apply: catch the latest labeled release before 2010-12-20 and append `+date=2010-12-20` to it.

Reply:
I wanted to clear up how we want to convert the current unstable format. We don't know the latest release and manually researching that for every package sounds not like something which I want to do for thousands of packages. Maybe we should do a special exception for converting packages? How about we do 0+date=2010-12-20 for converted packages but explicitly say that this is not how it should be for new or updated packages.

Also would it be easy to update the unstable updater to this? I could imagine that this could be quite hard. If we can't achieve that we should maybe change this to make our lives easier.

Reply:
I am not sure we want to commit to instant conversion. I think «that's what you do when adding/bumping packages» could make mass-conversion a more feasible amount of work once enough packages get converted during bumping.

As long as something is not touched, we can wait a bit longer than for high-churn packages.

Reply:
> that's what you do when adding/bumping packages

I am a bit less optimistic than probably most people here. If we take a look at the pname+version conversion https://github.com/NixOS/nixpkgs/issues/103997 we are still converting hundreds of packages despite that being a year old.